### ISSUE TITLE
Upgrade headers break HTTP to HTTPS redirects

### ISSUE THREAD

**Issue by rslife** [sagebind/isahc#159]
``` rust
use isahc::prelude::*;
use isahc::config::RedirectPolicy;

fn main() {
    //let req = Request::get("https://yahoo.com")
    let req = Request::get("http://yahoo.com")
        .redirect_policy(RedirectPolicy::Follow)
        .body(())
        .unwrap();

    let resp = req
        .send()
        .unwrap();

    eprintln!("{:?}", resp.effective_uri());
    eprintln!("{:?}", resp);
}
```

In a case I can't share, a 302 caused NoResponse errors which was confusing.

Issue does not exist in the curl crate (`Easy` and `Easy2` tested).

**Comment by rslife** [comment]
Okay. Upgrade headers seem to be causing the issue. Setting version negotiation to `VersionNegotiation::http11()` fixes it.

**Comment by sagebind** [comment]
What sort of response is the server returning for the initial request? This behavior sounds like a broken server implementation of HTTP/2 upgrading to me. I'll still keep investigating to see if I can understand better what might be happening, though it would help if I had a better idea of what sort of flow is happening. Here's kind of what I would expect:

1. Client requests `http://foo.com` with `Upgrade: h2c`
2. Server responds with an HTTP/1.1 302 pointing to `https://foo.com`
3. Client connects to `https://foo.com`
4. Server negotiates HTTP/2 via ALPN
5. Client performs request via HTTP/2

Or, if the server supports insecure HTTP/2:

1. Client requests `http://foo.com` with `Upgrade: h2c`
2. Server responds with an HTTP/1.1 101, then responds with an HTTP/2 302 pointing to `https://foo.com`
3. Client connects to `https://foo.com`
4. Server negotiates HTTP/2 via ALPN
5. Client performs request via HTTP/2

Bare curl would work because it does not attempt to use HTTP/2 by default, while Isahc does if possible.

**Comment by rslife** [comment]
There is a [reason](https://curl.haxx.se/mail/lib-2015-12/0054.html) why curl/libcurl does not send those headers by default. They tend to break some HTTP/1.1 servers (It's not a simple case of a broken HTTTP/2 server).

The reported servers in the case linked and yahoo give 400 errors. The private case that I ran into the responses became empty.

Sending those headers shouldn't be the default. It's as simple as that.

**Comment by sagebind** [comment]
Thanks for your detailed response. That's unfortunate that so many servers respond in this way, I will tweak the default behavior. I think a safe default then would be to only try to negotiate HTTP/2 via ALPN/TLS. I think _some_ level of HTTP/2 negotiation by default is warranted (as long as it is safe to do). What do you think?

**Comment by rslife** [comment]
> I think a safe default then would be to only try to negotiate HTTP/2 via ALPN/TLS.

Of course. That goes without saying. The issues are with http, not https.

**Comment by sagebind** [comment]
This change is now available in the [0.9](https://github.com/sagebind/isahc/releases/tag/0.9.0) release!
